### ISSUE TITLE
How can I save a model's checkpoint every N optimization steps?

### ISSUE THREAD

**Issue by jonatasgrosman** [PyTorchLightning/pytorch-lightning#5473]
Hello guys!

I'm trying to train a model with a really huge dataset that requires a lot of steps to complete an epoch (indeed, I'll probably train this model for just one or two epochs), and I'll need to save a model's checkpoint every N optimization steps.

I know that I can save the checkpoint every epoch using the [ModelCheckpoint](https://pytorch-lightning.readthedocs.io/en/stable/generated/pytorch_lightning.callbacks.ModelCheckpoint.html#pytorch_lightning.callbacks.ModelCheckpoint) callback or at the end of training using [trainer.save_checkpoint](https://pytorch-lightning.readthedocs.io/en/stable/weights_loading.html?highlight=trainer.save_checkpoint#manual-saving), but how can I save the checkpoint every 100k steps, for example?

**Comment by rohitgr7** [comment]
either you can use [val_check_interval](https://pytorch-lightning.readthedocs.io/en/latest/trainer.html#val-check-interval) or if you don't' have val_data, then you can use
```python
def training_step(...):
    ...
    if self.global_step % 100_000 == 0:
        self.trainer.save_checkpoint()
```
for this you might not need `ModelCheckpoint` at all I guess since `save_checkpoint` won't monitor any metric, it just creates a checkpoint.

**Comment by jonatasgrosman** [comment]
Thanks, @rohitgr7 I'll try this!

**Comment by indigoviolet** [comment]
For future reference: here's a slightly (imo) nicer callback version that also uses consistent paths:

```python
import pytorch_lightning as pl
from pytorch_lightning.callbacks import ModelCheckpoint

class PeriodicCheckpoint(ModelCheckpoint):
    def __init__(self, every: int):
        super().__init__(period=-1)
        self.every = every

    def on_train_batch_end(
        self, trainer: pl.Trainer, pl_module: pl.LightningModule, *args, **kwargs
    ):
        if pl_module.global_step % self.every == 0:
            assert self.dirpath is not None
            current = Path(self.dirpath) / f"latest-{pl_module.global_step}.ckpt"
            prev = (
                Path(self.dirpath) / f"latest-{pl_module.global_step - self.every}.ckpt"
            )
            trainer.save_checkpoint(current)
            prev.unlink(missing_ok=True)
```

**Comment by jonatasgrosman** [comment]
@indigoviolet, I think your approach is better indeed. Thanks! 😃

**Comment by rohitgr7** [comment]
you can also extend the above to allow additional kwargs for `ModelCheckpoint` like `dirpath`, `filename`, ...
